### PR TITLE
updated to use pivot_longer and pivot_wider functions

### PR DESCRIPTION
**Pull request by william3031**
Updated using the new functions.

### CODE REVIEW

File: vignettes/tidy-data.Rmd [hunk 1]
```diff
@@ -54,9 +54,10 @@ A tidy version of the pregnancy data looks like this: (you'll learn how the func
 ```{r, message = FALSE}
 library(tidyr)
 library(dplyr)
-preg2 <- preg %>% 
-  gather(treatment, n, treatmenta:treatmentb) %>%
-  mutate(treatment = gsub("treatment", "", treatment)) %>%
+library(stringr)
+preg2 <-preg %>% 
+  pivot_longer(cols = c(treatmenta:treatmentb), names_to = "treatment", values_to = "n") %>%
```
Comment (hadley) [inline vignettes/tidy-data.Rmd]: Can you please omit the name of the `cols` argument?

File: vignettes/tidy-data.Rmd [hunk 2]
```diff
@@ -54,9 +54,10 @@ A tidy version of the pregnancy data looks like this: (you'll learn how the func
 ```{r, message = FALSE}
 library(tidyr)
 library(dplyr)
-preg2 <- preg %>% 
-  gather(treatment, n, treatmenta:treatmentb) %>%
-  mutate(treatment = gsub("treatment", "", treatment)) %>%
+library(stringr)
+preg2 <-preg %>%
```
Comment (hadley) [inline vignettes/tidy-data.Rmd]: Missing a space here

File: vignettes/tidy-data.Rmd [hunk 3]
```diff
@@ -137,15 +138,17 @@ This format is also used to record regularly spaced observations over time. For
 billboard
 ```
 
-To tidy this dataset, we first gather together all the `wk` columns. The column names give the `week` and the values are the `rank`s:
+To tidy this dataset, we first use pivot_longer to gether up all the `wk` columns. The column names give the `week` and the values are the `rank`s:
 
 ```{r}
 billboard2 <- billboard %>% 
-  gather(week, rank, wk1:wk76, na.rm = TRUE)
+  pivot_longer(cols = c(wk1:wk76), names_to = "week", values_to = "rank") %>%
```
Comment (hadley) [inline vignettes/tidy-data.Rmd]: And no need for `c()` here

File: vignettes/tidy-data.Rmd [hunk 4]
```diff
@@ -137,15 +138,17 @@ This format is also used to record regularly spaced observations over time. For
...
 billboard2 <- billboard %>% 
-  gather(week, rank, wk1:wk76, na.rm = TRUE)
+  pivot_longer(cols = c(wk1:wk76), names_to = "week", values_to = "rank") %>% 
+  filter(!is.na(rank)) %>% 
+  arrange(week, artist)
 billboard2
 ```
 
-Here we use `na.rm` to drop any missing values from the gather columns. In this data, missing values represent weeks that the song wasn't in the charts, so can be safely dropped.
+Here we use `filter(!is.na())` to drop any missing values from the rank column. In this data, missing values represent weeks that the song wasn't in the charts, so can be safely dropped.
```
Comment (hadley) [inline vignettes/tidy-data.Rmd]: Can you please use `values_drop_na = TRUE` instead?

File: vignettes/tidy-data.Rmd [hunk 5]
```diff
@@ -172,18 +175,20 @@ billboard3 %>% arrange(date, rank)
...
-After gathering columns, the key column is sometimes a combination of multiple underlying variable names. This happens in the `tb` (tuberculosis) dataset, shown below. This dataset comes from the World Health Organisation, and records the counts of confirmed tuberculosis cases by `country`, `year`, and demographic group. The demographic groups are broken down by `sex` (m, f) and `age` (0-14, 15-25, 25-34, 35-44, 45-54, 55-64, unknown).
+After pivoting columns, the key column is sometimes a combination of multiple underlying variable names. This happens in the `tb` (tuberculosis) dataset, shown below. This dataset comes from the World Health Organisation, and records the counts of confirmed tuberculosis cases by `country`, `year`, and demographic group. The demographic groups are broken down by `sex` (m, f) and `age` (0-14, 15-25, 25-34, 35-44, 45-54, 55-64, unknown).
 
 ```{r}
 tb <- as_tibble(read.csv("tb.csv", stringsAsFactors = FALSE))
 tb
 ```
 
-First we gather up the non-variable columns:
+First we use pivot_longer to gather up the non-variable columns:
```
Comment (hadley) [inline vignettes/tidy-data.Rmd]:
```suggestion
First we use `pivot_longer()` to gather up the non-variable columns:
```

File: vignettes/tidy-data.Rmd [hunk 6]
```diff
@@ -172,18 +175,20 @@ billboard3 %>% arrange(date, rank)
...
 
 ```{r}
 tb2 <- tb %>% 
-  gather(demo, n, -iso2, -year, na.rm = TRUE)
+  pivot_longer(cols = c(-iso2, -year), names_to = "demo", values_to = "n") %>% 
+  filter(!is.na(n)) %>%
```
Comment (hadley) [inline vignettes/tidy-data.Rmd]: And here

File: vignettes/tidy-data.Rmd [hunk 7]
```diff
@@ -228,10 +234,11 @@ weather3 <- weather2 %>%
 weather3
 ```
 
-This dataset is mostly tidy, but the `element` column is not a variable; it stores the names of variables. (Not shown in this example are the other meteorological variables `prcp` (precipitation) and `snow` (snowfall)). Fixing this requires the spread operation. This performs the inverse of gathering by spreading the `element` and `value` columns back out into the columns:
+This dataset is mostly tidy, but the `element` column is not a variable; it stores the names of variables. (Not shown in this example are the other meteorological variables `prcp` (precipitation) and `snow` (snowfall)). Fixing this requires the pivot_wider operation. This performs the inverse of pivot_longer by spreading the `element` and `value` columns back out into the columns:
```
Comment (hadley) [inline vignettes/tidy-data.Rmd]:
```suggestion
This dataset is mostly tidy, but the `element` column is not a variable; it stores the names of variables. (Not shown in this example are the other meteorological variables `prcp` (precipitation) and `snow` (snowfall)). Fixing this requires widening the data. This performs the inverse of `pivot_longer()` by pivoting the `element` and `value` columns back out into the columns:
```

File: vignettes/tidy-data.Rmd [hunk 8]
```diff
@@ -228,10 +234,11 @@ weather3 <- weather2 %>%
...
 
 ```{r}
-weather3 %>% spread(element, value)
+weather3 %>% 
+  pivot_wider(names_from = "element", values_from = value)
```
Comment (hadley) [inline vignettes/tidy-data.Rmd]:
```suggestion
  pivot_wider(names_from = element, values_from = value)
```

File: vignettes/tidy-data.Rmd (hunk 1 above)
Comment (william3031) [reply vignettes/tidy-data.Rmd]: I assume that you mean `pivot_longer((treatmenta:treatmentb), names_to = "treatment", values_to = "n")`, if so, then done.

File: vignettes/tidy-data.Rmd (hunk 2 above)
Comment (william3031) [reply vignettes/tidy-data.Rmd]: Fixed

File: vignettes/tidy-data.Rmd (hunk 3 above)
Comment (william3031) [reply vignettes/tidy-data.Rmd]: Removed

File: vignettes/tidy-data.Rmd (hunk 4 above)
Comment (william3031) [reply vignettes/tidy-data.Rmd]: Sure. Done here and elsewhere.

File: vignettes/tidy-data.Rmd (hunk 5 above)
Comment (william3031) [reply vignettes/tidy-data.Rmd]: Changed.

File: vignettes/tidy-data.Rmd (hunk 6 above)
Comment (william3031) [reply vignettes/tidy-data.Rmd]: Here too.

**Comment by hadley** [comment]
Thanks!

**Comment by william3031** [comment]
Should I return it to `gsub`? I have only used one `stringr` function (`str_remove`).

**Comment by jennybc** [comment]
> Should I return it to `gsub()`? I have only used one stringr function (`str_remove`).

Yeah, I think that is for the best. This is a gotcha with vignettes: you really do have to be conscious of introducing new dependencies. And stringr brings in stringi, which is definitely not one to take on lightly.

**Pull request merged**